contribute a fix_data.yaml file to enable 'dart fix' renames

devoncarew opened the pull request:
- contribute a fix_data.yaml file to enable 'dart fix' renames
- document the class renames from `dart:html` to `package:web`
- related to https://github.com/dart-lang/web/issues/139

@keertip - can you review the fix_data.yaml file? I have some questions about it (in-line).
cc @bwilkerson and @jacob314 as we discussed this yesterday

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

devoncarew commented on example/example.dart:
```diff
@@ -5,6 +5,11 @@
 void main() {
-  final div = document.querySelector('div') as HTMLDivElement;
+  // ignore: unused_local_variable
+  final HtmlDocument foo = document;
```
This location _does_ offer a quick fix, to rename to `Document`.

devoncarew commented on example/example.dart:
```diff
@@ -5,6 +5,11 @@
+  // ignore: unused_local_variable
+  final HtmlDocument foo = document;
+
+  final doc = document as HtmlDocument;
```
This location does _not_ offer a quick fix. Should I define the fix data differently? Is the rename fix expected to cover this case?

cc @keertip

bwilkerson replied:
I would have expected the rename change to cover this case. It looks like a bug to me.

keertip replied:
As Brian said, it should, otherwise it is a bug. Could you open an issue to capture this?

devoncarew replied:
https://github.com/dart-lang/sdk/issues/54617

devoncarew commented:
Somewhat related - and I can open a separate issue for this - quick fix will currently un-do the work of the `dart fix` fixes. When I open the quick fixes available at a ctor reference to an older symbol from dart:html (w/ just package:web in the imports, not dart:html):

<img width="444" alt="Screenshot 2024-01-12 at 8 44 54 AM" src="https://github.com/dart-lang/web/assets/1269969/ca8da234-a267-483d-8152-a0d3b4bf6843">

The 3rd item above is what we want the user to select - to rename the older symbol name to the new (`HTMLDivElement`) name. However the first item suggested - adding an import for dart:html - will actually work against the migration, and will likely be confusing to casual users. We should probably stop suggesting adding imports for the legacy web libraries (eg, https://api.dart.dev/main/a5758a0dc373e137ee2cf486b5a537fbd5696a95/index.html).

devoncarew commented:
It would be great to be able to bump the 'rename' quick fix to be first in the list in any case - for it to be suggested ahead of the 'create new class' fix.

parlough commented:
> and will likely be confusing to casual users. We should probably stop suggesting adding imports for the legacy web libraries

+1000 This is also what I think would be most useful in the migration. Something like https://github.com/dart-lang/sdk/issues/54347

bwilkerson commented on lib/fix_data.yaml:
```diff
@@ -0,0 +1,1342 @@
+  - title: "Rename to 'HTMLAnchorElement'"
+    date: 2024-01-12
+    element:
+      uris: [ 'package:web/web.dart' ]
```
Either ought to be fine. The shorter form is an "abbreviated URI". See https://github.com/flutter/flutter/wiki/Data-driven-Fixes#uri.

keertip replied:
Either should be fine. Though we do have this open issue - https://github.com/dart-lang/sdk/issues/52233.

devoncarew replied:
Hmm - is that issue a blocker? Does it mean we should include both types of urls in the fix data file (`web.dart` and `package:web/web.dart`)? Is this is going to affect us it would be great to prioritize that fix for the next release.

bwilkerson reviewed (commented):
Consider defining tests for each of the transformations as per https://github.com/flutter/flutter/wiki/Data-driven-Fixes#testing. I strongly recommend having at least one `.dart` file per transform, even though the flutter team has chosen to put the tests for multiple transforms in a single file. I think it makes it much easier to debug problems when the test files are smaller.

devoncarew commented:
> Consider defining tests for each of the transformations as per https://github.com/flutter/flutter/wiki/Data-driven-Fixes#testing.

Thanks for the feedback and the idea! I added content in a `test_fixes/` dir. It did find 1/2 dozen quick fixes that were not taking effect - places where we had introduced typedefs to help w/ migrations.

bwilkerson commented:
Glad we found the bugs.

srujzs reviewed (approved):
Awesome, thanks for working on this! I commented on the other issue on why we might be missing some of the mappings. (I'm approving from my perspective - better to have some of the analyzer folks approve before submitting)

keertip reviewed (approved):
LGTM for the fix_data.yaml file. Please do test that dart fix works if you change the uri from 'package:'.

devoncarew commented:
> LGTM for the fix_data.yaml file. Please do test that dart fix works if you change the uri from 'package:'.

I did try updating the fix_data references to use `web.dart` (from `package:web/web.dart`). The fixes didn't work after that - either from the editor (in the example file) or via the cli testing (`--compare-to-golden`) technique. Keeping the longer form for now.
